velox/exec/TopNRowNumber.cpp

```diff
@@ -249,25 +249,52 @@ void TopNRowNumber::initializeNewPartitions() {
   }
 }
 
+char* TopNRowNumber::processRowBeforeLimit(
+    vector_size_t /*index*/,
+    TopRows& partition) {
+  // row_number accumulates the new row in the partition, and the top rank is
+  // incremented by 1 as row_number increases by 1 at each new row.
+  partition.topRank++;
+  return data_->newRow();
+}
+
+char* TopNRowNumber::processRowBeyondLimit(
+    vector_size_t /*index*/,
+    TopRows& partition) {
+  // The new row has rank < highest (aka top) rank at 'limit' function value.
+  // For row_number, such rows are added to the accumulator queue and the
+  // top rank row is popped out. The topRank remains the same.
+  auto& topRows = partition.rows;
+  char* topRow = topRows.top();
+  topRows.pop();
+  // Reuses the space of the popped row itself for the new row.
+  return data_->initializeRow(topRow, true /* reuse */);
+}
+
 void TopNRowNumber::processInputRow(vector_size_t index, TopRows& partition) {
   auto& topRows = partition.rows;
 
   char* newRow = nullptr;
-  if (topRows.size() < limit_) {
-    newRow = data_->newRow();
+  if (partition.topRank < limit_) {
+    newRow = processRowBeforeLimit(index, partition);
   } else {
     char* topRow = topRows.top();
 
-    if (!comparator_(decodedVectors_, index, topRow)) {
-      // Drop this input row.
+    auto result = comparator_.compare(decodedVectors_, index, topRow);
+    if (result < 0) {
```

Comment:
Can we combine these conditions as following?
```
if (result <= 0) {
return;
} else {
newRow = processRowBeyondLimit(index, partition);
}
```

Reply:
Same as below. rank and dense_rank have some difference in behavior, so retaining it to avoid further diffs.